client/rest/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -166,73 +167,159 @@ public Response performRequest(String method, String endpoint, Map<String, Strin
+                               ResponseListener responseListener, Header... headers) {
+        performRequest(method, endpoint, params, null, responseListener, headers);
+    }
+
+    /**
+     /**
+     * Sends a request to the elasticsearch cluster that the current client points to.
+     * Shortcut to {@link #performRequest(String, String, Map, HttpEntity, HttpAsyncResponseConsumer, ResponseListener, Header...)}
+     * but without an async response consumer, meaning that a {@link org.apache.http.nio.protocol.BasicAsyncResponseConsumer}
+     * will be created and used.
+     *
+     * @param method the http method
+     * @param endpoint the path of the request (without host and port)
+     * @param params the query_string parameters
+     * @param entity the body of the request, null if not applicable
+     * @param responseListener the {@link ResponseListener} to notify when the request is completed or fails
+     * @param headers the optional request headers
+     */
+    public void performRequest(String method, String endpoint, Map<String, String> params,
+                               HttpEntity entity, ResponseListener responseListener, Header... headers) {
+        HttpAsyncResponseConsumer<HttpResponse> responseConsumer = HttpAsyncMethods.createConsumer();
+        performRequest(method, endpoint, params, entity, responseConsumer, responseListener, headers);
+    }
+
+    /**
+     * Sends a request to the elasticsearch cluster that the current client points to. The request is executed asynchronously
+     * and the provided {@link ResponseListener} gets notified whenever it is completed or it fails.
+     * Selects a host out of the provided ones in a round-robin fashion. Failing hosts are marked dead and retried after a certain
+     * amount of time (minimum 1 minute, maximum 30 minutes), depending on how many times they previously failed (the more failures,
+     * the later they will be retried). In case of failures all of the alive nodes (or dead nodes that deserve a retry) are retried
+     * till one responds or none of them does, in which case an {@link IOException} will be thrown.
+     *
+     * @param method the http method
+     * @param endpoint the path of the request (without host and port)
+     * @param params the query_string parameters
+     * @param entity the body of the request, null if not applicable
+     * @param responseConsumer the {@link HttpAsyncResponseConsumer} callback
+     * @param responseListener the {@link ResponseListener} to notify when the request is completed or fails
+     * @param headers the optional request headers
+     */
+    public void performRequest(String method, String endpoint, Map<String, String> params,
+                               HttpEntity entity, HttpAsyncResponseConsumer<HttpResponse> responseConsumer,
+                               ResponseListener responseListener, Header... headers) {
```

Comment:
I played quite a bit with this and gave it quite some thoughts which I'd like to sum up here.

What the sync client provides is effectively a way to hold on to the connection and read the incoming content directly through `InputStream` (`response.getEntity().getContent()`). That is why you then get a `CloseableHttpResponse` and the connection is not released if you forget to close it.

What I just described is possible with a blocking approach, but not the same with a non blocking approach. With the move to the async client, the work of reading bytes is done by the `HttpAsyncResponseConsumer`, the client side cannot block the connection. The default consumer implementation buffers the bytes of the entire response in heap. so when you get an actual response through the listener, it is not a closeable one and the whole response body is already entirely read. You read from an input stream still, but that one reads from the heap buffer, connection is long closed already.

This is a reasonable compromise when wanting to do non blocking IO, but is quite a shame that people wanting to send sync requests still end up buffering the whole response in memory, as we are moving entirely to the async client internally (that means no `CloseableHttpResponse` for the sync `performRequest` methods too!).

I looked into whether and how we could bridge the two worlds to address this but figured that it would make things very complex. Also, most of our users are still going to parse the response and buffer it in memory anyways later. Parsers usually accept an InputStream to parse from, which is something not easy to build from an `HttpAsyncResponseConsumer`. I do not think allowing users to parse directly from a response consumer even makes sense nor is possible. Maybe the buffering is not a huge deal. Also, the heap buffer may become a problem only with a minority of usecases, which we shouldn't try to address and solve just right now.

For the above reasons, I did the following:

- add our own response consumer (`HeapBufferAsyncResponseConsumer`) that still buffers the whole response in heap, but has a much lower limit, 10MB rather than Integer.MAX_VALUE (2GB). Beyond that size it will throw a `ContentTooLongException` and users need to provide their own response consumer, which they can write from scratch, or use our own and increase the buffer limit explicitly (through the bufferLimit constructor argument)

- add a sync variant of performRequest that also accepts an instance of `HttpAsyncResponseConsumer`

most of our users will buffer the whole response in memory and have no problems with that. Advanced users can provide their own implementation of `HttpAsyncResponseConsumer`, where they can effectively even read the response body [byte by byte](http://hc.apache.org/httpcomponents-asyncclient-dev/httpasyncclient/apidocs/org/apache/http/nio/client/methods/AsyncByteConsumer.html) or [char by char](http://hc.apache.org/httpcomponents-asyncclient-dev/httpasyncclient/apidocs/org/apache/http/nio/client/methods/AsyncCharConsumer.html).

For simplicity I decided to stick with `HttpResponse`  ( `HttpAsyncResponseConsumer<HttpResponse>` that is) rather than some generic or our own object, cause that would complicate things quite a bit. People wanting to do things very differently can still write their own `HttpResponse` and their own `HttpEntity` and plug it in through their own `HttpAsyncResponseConsumer` if needed (although I am not sure this will happen). This approach has also the benefit of being just some async http client customization rather than something specific for our java REST client.